runtime/native/jvm.cpp

```diff
@@ -625,14 +627,143 @@ JVM_GetClassAnnotations(JNIEnv *env, jclass cls) {
+    void Polyglot_java_lang_reflect_Method_Method__Ljava_lang_Class_2Ljava_lang_String_2_3Ljava_lang_Class_2Ljava_lang_Class_2_3Ljava_lang_Class_2IILjava_lang_String_2_3B_3B_3B (jobject, jclass, jstring, jobjectArray, jclass, jobjectArray, jint, jint, jstring, jbyteArray, jbyteArray, jbyteArray);
+}
+
 jobjectArray
 JVM_GetClassDeclaredMethods(JNIEnv *env, jclass ofClass, jboolean publicOnly) {
-    JvmUnimplemented("JVM_GetClassDeclaredMethods");
+    const JavaClassInfo* info = GetJavaClassInfo(ofClass);
+    if (info) {
+        // if primative class (int, boolean), return empty array
+        if (JVM_IsPrimitiveClass(env, ofClass)) {
+            return CreateJavaObjectArray(0);
+        }
+
+        jclass MethodClass = env->FindClass("java.lang.reflect.Method");
+
+        // TODO take into account publiconly argument
+
+        jobjectArray ret = CreateJavaObjectArray(info->num_methods);
+
+        JavaMethodInfo* methods = info->methods;
+
+        for (int i = 0; i < info->num_methods; i++) {
+            // create new field object
+            jobject newMethod = CreateJavaObject(MethodClass);
+            jstring nameString = env->NewStringUTF(methods[i].name);
+            // TODO need to get the proper values
+            // call the method constructor
+            Polyglot_java_lang_reflect_Method_Method__Ljava_lang_Class_2Ljava_lang_String_2_3Ljava_lang_Class_2Ljava_lang_Class_2_3Ljava_lang_Class_2IILjava_lang_String_2_3B_3B_3B(newMethod, NULL, nameString, NULL, NULL, NULL, 0, 0, NULL, NULL, NULL, NULL);
+            // add it to the array
+            JVM_SetArrayElement(env, ret, i, newMethod);
+        }
+
+        return ret;
+    }
+    return NULL;
 }
 
 jobjectArray
 JVM_GetClassDeclaredFields(JNIEnv *env, jclass ofClass, jboolean publicOnly) {
-    JvmUnimplemented("JVM_GetClassDeclaredFields");
+    const JavaClassInfo* info = GetJavaClassInfo(ofClass);
+    if (info) {
+        // if primative class (int, boolean), return empty array
+        if (JVM_IsPrimitiveClass(env, ofClass)) {
+            return CreateJavaObjectArray(0);
+        }
+
+        jclass FieldsClass = env->FindClass("java.lang.reflect.Field");
+        // jclass ClassClass = env->FindClass("java.lang.Class");
+        // const JavaClassInfo* classInfo = GetJavaClassInfo(ClassClass);
+        // printf("class size: %d\n", classInfo->obj_size);
+
+        // dw475 TODO take into account publiconly argument
+        // dw475 TODO take super fields into account
+
+        // non-static and static fields
+        jobjectArray ret = CreateJavaObjectArray(info->num_fields + info->num_static_fields);
+
+        JavaFieldInfo* fields = info->fields;
+        JavaStaticFieldInfo* staticFields = info->static_fields;
+
+        for (int i = 0; i < info->num_fields+info->num_static_fields; i++) {
+            // create new field object
+            jobject newField = CreateJavaObject(FieldsClass);
+
+            // // using a Field Struct
```

Comment:
This large commented section is to show how this would be done without calling in to Java code right?
Maybe elaborate on the 'using a Field Struct' comment to make that clearer.

Maybe even add an issue related to performance which says that we should do evaluations of these two approaches.